sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,320 @@
+    }
+
+    @Override
+    public void validate() {
+      boundedSource.validate();
+    }
+
+    @Override
+    public List<BoundedToUnboundedSourceAdapter<T>> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      long desiredBundleSize = boundedSource.getEstimatedSizeBytes(options) / desiredNumSplits;
+      List<BoundedToUnboundedSourceAdapter<T>> result = Lists.newArrayList();
+      for (BoundedSource<T> split : boundedSource.splitIntoBundles(desiredBundleSize, options)) {
+        result.add(new BoundedToUnboundedSourceAdapter<>(split));
+      }
+      return result;
+    }
+
+    @Override
+    public Reader createReader(PipelineOptions options, Checkpoint<T> checkpoint)
+        throws IOException {
+      if (checkpoint == null) {
+        return new Reader(null /* residualElements */, boundedSource, options);
+      } else {
+        return new Reader(checkpoint.residualElements, checkpoint.residualSource, options);
+      }
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return boundedSource.getDefaultOutputCoder();
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public Coder<Checkpoint<T>> getCheckpointMarkCoder() {
+      return new CheckpointCoder<>(boundedSource.getDefaultOutputCoder());
+    }
+
+    @VisibleForTesting
+    static class Checkpoint<T> implements UnboundedSource.CheckpointMark {
+      private final List<TimestampedValue<T>> residualElements;
+      private final BoundedSource<T> residualSource;
+
+      public Checkpoint(
+          List<TimestampedValue<T>> residualElements, BoundedSource<T> residualSource) {
+        this.residualElements = residualElements;
+        this.residualSource = residualSource;
+      }
+
+      @Override
+      public void finalizeCheckpoint() {}
+    }
+
+    private static class CheckpointCoder<T> extends AtomicCoder<Checkpoint<T>> {
+      private final Coder<List<TimestampedValue<T>>> elemsCoder;
+      private final Coder<T> elemCoder;
+
+      @JsonCreator
+      public static CheckpointCoder<?> of(
+          @JsonProperty(PropertyNames.COMPONENT_ENCODINGS)
+          List<Coder<?>> components) {
+        Preconditions.checkArgument(components.size() == 1,
+            "Expecting 1 components, got " + components.size());
+        return new CheckpointCoder<>(components.get(0));
+      }
+
+      @SuppressWarnings("rawtypes")
+      private final SerializableCoder<BoundedSource> serializableCoder =
+          SerializableCoder.of(BoundedSource.class);
+
+      CheckpointCoder(Coder<T> elemCoder) {
+        this.elemsCoder = ListCoder.of(TimestampedValue.TimestampedValueCoder.of(elemCoder));
+        this.elemCoder = elemCoder;
+      }
+
+      @Override
+      public void encode(Checkpoint<T> value, OutputStream outStream, Context context)
+          throws CoderException, IOException {
+        elemsCoder.encode(value.residualElements, outStream, context);
```

Comment:
This should be context.nested(); the residual list will never be the entire stream